### PR TITLE
ci: Auto-update gemfile versions

### PR DESCRIPTION
**Pull request by quinna-h**
**What does this PR do?**

**Motivation:**

**Additional Notes:**

**How to test the change?**

Unsure? Have a question? Request a review!

### CODE REVIEW

**Comment by bouwkast** [comment]
:smile: I didn't realize that there was a local code quality step sorry

File: .github/workflows/update-gemfiles.yml [hunk 1]
```diff
@@ -3,78 +3,57 @@ name: Update Gemfiles
 # This action cannot be skipped altogether because it is a mandatory status check.
 # Instead we conditionally skip it at job level, instead of workflow level.
 on:
-  # Execute on `push` and not `pull_request` because `pull_request`
```
Comment (bouwkast) [inline .github/workflows/update-gemfiles.yml:6]: Ahh interesting, I wasn't aware of this workflow - would this have been getting called and regenerating all of the `gemfile/` directory?

File: .github/scripts/version_updater.rb [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+                gem_version = dep.requirements_list
+                gem_requirements = Gem::Requirement.create(*gem_version)
+                gem_platforms = dep.platforms
+        
+                puts "Gem Name: #{gem_name}"
+                puts "Gem Version: #{gem_version}"
+                puts "Gem Requirments: #{gem_requirements}"
+                puts "Gem Platforms: #{gem_platforms}"
+        
+                latest = Gem.latest_version_for(gem_name)
```
Comment (bouwkast) [inline .github/scripts/version_updater.rb:49]: This `latest = Gem.latest_version_for(gem_name)` can actually be moved outside of this loop (should save some time)

```suggestion
        
            latest = Gem.latest_version_for(gem_name)
            dependencies.each do |dep|
                next if gem_to_update != dep.name # only update the gem we chose at random
                # puts "Skipping #{dep.name} as it is not #{gem_to_update}"
                
                gem_name = dep.name
                gem_version = dep.requirements_list
                gem_requirements = Gem::Requirement.create(*gem_version)
                gem_platforms = dep.platforms
        
                puts "Gem Name: #{gem_name}"
                puts "Gem Version: #{gem_version}"
                puts "Gem Requirments: #{gem_requirements}"
                puts "Gem Platforms: #{gem_platforms}"
```

File: .github/workflows/update-gemfiles.yml [hunk 3]
```diff
@@ -3,78 +3,62 @@ name: Update Gemfiles
...
-        name: Commit gemfiles changes, if any, back to the branch
-        uses: stefanzweifel/git-auto-commit-action@3ea6ae190baf489ba007f7c92608f33ce20ef04a # v4.16.0
+              - 'appraisal/**'
+              - 'gemfiles/**'
+
+      - name: Run update gemfiles if files changed
+        if: steps.filter.outputs.gemfile == 'true'
+        run: ruby .github/scripts/version_updater.rb
+
+      - name: Commit gemfiles changes, if any, back to the branch
```
Comment (quinna-h) [inline .github/workflows/update-gemfiles.yml:57]: This only creates the commit and pushes changes - need to use a different repo to create the PR

File: .github/scripts/version_updater.rb [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                    # For example gemfiles/ruby_2.5_contrib.gemfile.lock has "grpc (1.48.0-x86_64-linux)"
+                    # But this platform isn't specified in the gemfile
+                    # Run the command here removes this "grpc (1.48.0-x86_64-linux)" instance entirely
+                    # TODO check with Marco/Tony/David for how this came to be as the PR they were added in was a while back and was just an update
+
+                    puts "Updating #{gem_name} with: #{bundler_args}"
+                    run_bundler_command(gemfile, bundler_args)
+                else
+                    puts "The latest (#{latest}) of #{gem_name} is NOT satisfied by #{gem_requirements}"
+                    # TODO - should we attempt to do a --conservative update here?
```
Comment (quinna-h) [inline .github/scripts/version_updater.rb:66]: Do we still want to do this?

File: .github/scripts/version_updater.rb [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        paths = Dir['gemfiles/jruby_*.lock', 'gemfiles/ruby_*.lock']
+        
+        # TODO - a lock file can contain the same gem multiple times
+        # for example: grpc (1.66.0), grpc (1.66.0-aarch64-linux), and grpc (1.66.0-x86_64-linux)
+        # are all "grpc", but each one is "different", so the same gemfile in this loop gets added multiple times
+        # should probably fix this
+        parse_paths(paths)
+        # TODO - could we look through open PR titles to rule out which gems to update?
+        #        if the PR is the autogenerated PR to update the gem we could skip it
+        gem_to_update = integrated_gems.sample # HACKY pick a random gem to attempt to update
```
Comment (quinna-h) [inline .github/scripts/version_updater.rb:116]: Picking a random gem seems fine, maybe we want to be able to query github to see which PRs are already open and skip if a gem has already been added / updated

**Pull request closed without merging**